deps.ml

```diff
@@ -0,0 +1,141 @@
+     * we need to include the internal_deps in order to reach all the deps.
+     * If the internal dep is a public library we skip the recursive resolve
+     * because it will be resolve with separate request*)
+    let open Describe_external_lib in
+    let get_name = function
+      | Lib item  -> item.name ^ ".lib"
+      | Exe item  -> item.name ^ ".exe"
+      | Test item -> item.name ^ ".test"
+    in
+    let d_items_lib =
+      d_items
+      |> List.filter is_lib_item
+      |> List.map (fun d_item ->
+          d_item
+          |> get_item
+          |> (fun (item:Describe_external_lib.item) ->
+              (item.name ^ ".lib", Lib item)))
+      |> List.to_seq |> Hashtbl.of_seq
+    in
+    let rec add_internal acc = function
+      | [] -> Hashtbl.to_seq_values acc |> List.of_seq
+      | item::tl ->
+        if Hashtbl.mem acc (get_name item) then
+          add_internal acc tl
+        else begin
+          Hashtbl.add acc (get_name item) item;
+          (get_item item).internal_deps
+          |> List.filter (fun (_, k) -> Kind.is_required k)
+          |> List.filter_map (fun (name, _) ->
+              match Hashtbl.find_opt d_items_lib (name ^ ".lib") with
+              | None -> None
+              | Some d_item_lib ->
+                if Option.is_some (get_item d_item_lib).package then None else Some d_item_lib)
+          |> fun internals -> add_internal acc (tl @ internals)
+        end
+    in
+    add_internal (Hashtbl.create 10) items_pkg
+  in
+  describe_external_lib_deps
+  |> List.map (fun d_item ->
+      let item = Describe_external_lib.get_item d_item in
+      if Describe_external_lib.is_exe_item d_item && Option.is_none item.package
+      then
+        match find_exe_item_package item  with
+        | None ->  d_item
+        | Some pkg -> Describe_external_lib.Exe { item with package = Some pkg }
+      else d_item)
+  |> List.filter (fun item ->
+      match (item,target) with
+      | Describe_external_lib.Test _, `Install -> false
+      | Describe_external_lib.Test _, `Runtest -> true
+      | _ , `Runtest -> false
+      | _, `Install -> true)
+  |> List.filter (fun d_item -> should_use_dir ~dir_types (Describe_external_lib.get_item d_item).source_dir)
+  |> (fun d_items ->
+      d_items
+      |> List.filter (fun d_item ->
+          let item = Describe_external_lib.get_item d_item in
+          (* if an item has not package, we assume it's used for testing*)
+          if target = `Install then
+            Option.equal String.equal (Some pkg) item.package
+          else
+            Option.equal String.equal (Some pkg) item.package || Option.is_none item.package)
+      |> resolve_internal_deps d_items)
+
+
+let lib_deps ~pkg ~target =
+  get_dune_items (Hashtbl.create 10) ~pkg ~target
+  |> List.map Describe_external_lib.get_item
+  |> List.fold_left (fun acc (item:Describe_external_lib.item) ->
+      List.map (fun dep -> (fst dep, item.source_dir)) item.external_deps @ acc) []
```

Comment:
Not sure, but maybe you want `List.fold_left_map`?

Reply:
It is different, no need for `fold_left_map`.